packages/ui/src/views/run-pipeline/utils/inputs-utils.ts

```diff
@@ -73,26 +45,62 @@ export function pipelineInputs2FormInputs({
   return [...inputsFromLayout, ...remainingInputs]
 }
 
-export const validateUniqueInputKeysInLayout = (layout: InputLayout): string[] => {
-  const inputOccurrences = new Map<string, number>()
+/**
+ * Recursively processes layout into form inputs, skipping duplicate keys after their first appearance.
+ *
+ * @param layout - The layout definition
+ * @param pipelineInputs - The input data
+ * @param options - Options for input formatting
+ * @param processedInputKeys - Tracks already-added input keys
+ * @returns An array of input definitions
+ */
+const processLayout = (
+  layout: InputLayout,
+  pipelineInputs: Record<string, any>,
+  options: { prefix?: string },
+  processedInputKeys: Set<string>
+): IInputDefinition[] => {
+  return layout.flatMap(item => {
+    if (typeof item === 'string') {
+      if (processedInputKeys.has(item) || !(item in pipelineInputs)) return []
+      processedInputKeys.add(item)
+      return pipelineInput2FormInput(item, pipelineInputs[item], options)
+    }
 
-  const traverse = (layout: InputLayout) => {
-    layout.forEach(item => {
-      if (typeof item === 'string') {
-        inputOccurrences.set(item, (inputOccurrences.get(item) || 0) + 1)
-      } else {
-        traverse(item.items)
+    // If group has no title, flatten its items
+    if (!item.title && item.items && item.items.length > 0) {
+      return processLayout(item.items, pipelineInputs, options, processedInputKeys)
```

Comment:
Just a nit and maybe more of a preference, so not a big deal either way, but you could assign this line to a `const` before the `if` and return it here if we hit this branch else assign to `inputs` below if not.